Bottom: 5b2c797ca51cda3fae41277df909186373943956
Top:    3155b9b25baf3c5f0c94431c566a23537cdab5ea
Author: Jason Wang <jasowang@redhat.com>
Date:   2016-06-01 16:21:31 +0800

Refresh of tun-ex

---

diff --git a/drivers/net/tun.c b/drivers/net/tun.c
index e99bf78..2996157 100644
--- a/drivers/net/tun.c
+++ b/drivers/net/tun.c
@@ -1164,6 +1164,7 @@ static ssize_t tun_get_user(struct tun_struct *tun, struct tun_file *tfile,
 	u32 rxhash;
 	ssize_t n;
 	struct skb_msg *m = msg_control;
+	struct sk_buff_head *write_queue = &tfile->sk.sk_write_queue;
 
 	if (!(tun->dev->flags & IFF_UP))
 		return -EIO;
@@ -1332,7 +1333,23 @@ static ssize_t tun_get_user(struct tun_struct *tun, struct tun_file *tfile,
 	skb_probe_transport_header(skb, 0);
 
 	rxhash = skb_get_hash(skb);
-	netif_rx_ni(skb);
+
+	spin_lock(&write_queue->lock);
+	__skb_queue_tail(&tfile->sk.sk_write_queue, skb);
+
+	if (m && (m->flags & SKB_MSG_MORE)) {
+		skb = NULL;
+	} else {
+		struct sk_buff_head send;
+		__skb_queue_head_init(&send);
+		skb_queue_splice_tail_init(&tfile->sk.sk_write_queue, &send);
+		skb_peek_tail(&send)->next = NULL;
+		skb = skb_peek(&send);
+	}
+	spin_unlock(&write_queue->lock);
+
+	if (skb)
+		netif_rx_ni(skb);
 
 	stats = get_cpu_ptr(tun->pcpu_stats);
 	u64_stats_update_begin(&stats->syncp);
diff --git a/net/core/dev.c b/net/core/dev.c
index faeea81..4640654 100644
--- a/net/core/dev.c
+++ b/net/core/dev.c
@@ -3780,7 +3780,7 @@ drop:
 
 static int netif_rx_internal(struct sk_buff *skb)
 {
-	int ret;
+	int ret = 0;
 
 	while (skb) {
 		struct sk_buff *next = skb->next;
